cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -207,19 +207,19 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
 
 Result<std::shared_ptr<Tensor>> FixedShapeTensorType::MakeTensor(
     const std::shared_ptr<ExtensionScalar>& scalar) {
-  const auto ext_scalar = internal::checked_pointer_cast<ExtensionScalar>(scalar);
-  const auto ext_type =
-      internal::checked_pointer_cast<FixedShapeTensorType>(scalar->type);
+  const auto* ext_scalar = internal::checked_cast<const ExtensionScalar*>(scalar.get());
```

Comment:
Since we're enhancing the code style, I would recommend const references over const pointers where possible:
```suggestion
  const auto& ext_scalar = internal::checked_cast<const ExtensionScalar&>(*scalar.get);
```